compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -357,31 +357,16 @@ public void visit(BLangArrayLiteral arrayLiteral) {
             }
             actualType = new BArrayType(actualType, null, arrayLiteral.exprs.size(), BArrayState.UNSEALED);
 
-            if (expType.tag == TypeTags.UNION) {
-                BUnionType expType = (BUnionType) this.expType;
-                int count = 0;
-                // array literals can not be assigned to union types that include
-                // any[],
-                // union arrays eg:(int|boolean)[],
-                // two or more matching types eg: int[] | int[4]
-                for (BType memType : expType.memberTypes) {
-                    if (memType.tag == TypeTags.ARRAY) {
-                        if (((BArrayType) memType).eType.tag == TypeTags.ANY ||
-                                ((BArrayType) memType).eType.tag == TypeTags.UNION) {
-                            dlog.error(arrayLiteral.pos, DiagnosticCode.INVALID_ARRAY_LITERAL, expType);
-                            resultType = symTable.errType;
-                            return;
-                        }
-                        if (types.isAssignable(actualType, memType)) {
-                            checkExprs(arrayLiteral.exprs, this.env, ((BArrayType) memType).eType);
-                            count++;
-                        }
-                    }
-                }
-                if (count > 1) {
-                    dlog.error(arrayLiteral.pos, DiagnosticCode.INVALID_ARRAY_LITERAL, expType);
-                    resultType = symTable.errType;
-                    return;
+            List<BType> arrayCompatibleType = getArrayCompatibleTypes(expType, actualType);
+            if (arrayCompatibleType.stream().anyMatch(t -> t.tag == TypeTags.ANY)) {
```

Comment:
This check has to be done after Line 366. Also, this should give INVALID_ARRAY_LITERAL error.